Update encryption-at-rest task page (#33285)
kubernetes/website: content/en/docs/tasks/administer-cluster/encrypt-data.md

---
title: Encrypting Confidential Data at Rest
reviewers:
- smarterclayton
content_type: task
---

<!-- overview -->

Some of the APIs in Kubernetes, such as {{< glossary_tooltip text="Secret" term_id="secret" >}},
support at-rest encryption. This at-rest encryption is additional to any system-level

> encryption for the etcd cluster or hosts where the kube-apiserver stores data persistently.

Author: Actually I don't think this is about stacked etcd vs a separate cluster.

I think the point is that encryption at rest might also apply to, for example, audit logs written to the host filesystem by a static API server Pod.

If we make that change, then we're changing the meaning and would be leaving out that other scenario (logs, etc).

{{< glossary_tooltip text="volumes" term_id="volume" >}}
- encrypt the data within your own application
{{< /note >}}

## {{% heading "prerequisites" %}}

* {{< include "task-tutorial-prereqs.md" >}}

* Your cluster's control plane **must** use etcd v3.x (major version 3, any minor version)

<!-- steps -->

## Determining whether encryption at rest is already enabled

By default, the API server uses the `identity` provider to protect Secrets in etcd.
**The default `identity` provider does not provide any confidentiality protection.**

The `kube-apiserver` process accepts an argument `--encryption-provider-config`
that specifies a path to a configuration file. The contents of that file, if you specify one,
control how Kubernetes API data is encrypted in etcd.

The format of that configuration file is YAML, representing a configuration API kind named
[`EncryptionConfiguration`](/docs/reference/config-api/apiserver-encryption.v1/).
You can see an example configuration
in [Encryption at rest configuration](#understanding-the-encryption-at-rest-configuration).

Check whether the `kube-apiserver` process is running with the `--encryption-provider-config`

> command line argument is set. If it is not, you do not have encryption at rest is enabled.

Author: I think so, yes. If we can merge this as an improvement without rewriting, that's handy.

If `--encryption-provider-config` is set, check which resources (such as `secrets`) are
configured for encryption, and what provider is used.
Make sure that the preferred provider for that resource type is **not** `identity`; you
only set `identity` (_no encryption_) as default when you want to disable encryption at
rest.
Then the first-listed provider for a resource is something **other** than `identity`,
then any new information written to resources of that type will be encrypted as configured.

> If you are not sure about the progress of any previous migration to encrypt data at rest,

Author: Are you able to recommend a different place to put this advice @tengqm ? If so, I can move it there.

> If you are not sure about the progress of any previous migration to encrypt data at rest,

Reviewer: Maybe this warrants a dedicated subsection, "migrating to encryption at rest".

you must overwrite each resource with a new updated copy, in order to trigger encryption.
See [Encrypting your data](#encrypting-your-data) for details.

## Encryption-at-rest configuration {#understanding-the-encryption-at-rest-configuration}

{{< caution >}}
For cluster configurations with two or more control plane nodes, the encryption configuration
**must** be identical across each control plane node.

If there is a difference in the encryption provider configuration, this may well mean
that the kube-apiserver can't decrypt data stored inside the key-value store (potentially
leading to further problems).
{{< /caution >}}

Here is an example EncryptionConfiguration file for the kube-apiserver:

```yaml
---
#
# CAUTION: this is an example configuration.
#          Do not use this for your own cluster!
#
apiVersion: apiserver.config.k8s.io/v1
kind: EncryptionConfiguration
resources:
  - resources:
      - secrets
    providers:

>       - identity: {} # default to NO encryption

Author: The fact that `identity` is default does not seem relevant here. Should we emphasize that detail?

* To encrypt a custom resource, your cluster must be running Kubernetes v1.26 or newer.
  {{< version-check >}}

<!-- steps -->

## Determine whether encryption at rest is already enabled {#determining-whether-encryption-at-rest-is-already-enabled}

By default, the API server uses an `identity` provider to protect confidential data
within etcd.
**The default `identity` provider does not provide any confidentiality protection.**

The `kube-apiserver` process accepts an argument `--encryption-provider-config`
that specifies a path to a configuration file. The contents of that file, if you specify one,
control how Kubernetes API data is encrypted in etcd.

The format of that configuration file is YAML, representing a configuration API kind named
[`EncryptionConfiguration`](/docs/reference/config-api/apiserver-encryption.v1/).
You can see an example configuration
in [Encryption at rest configuration](#understanding-the-encryption-at-rest-configuration).

Check whether the `kube-apiserver` process is running with the `--encryption-provider-config`
command line argument set to `true`. If it is not, you do not have encryption at rest enabled.

> Otherwise, you need to perform further checks to confirm the situation.

Reviewer: line 64 is confusing... what am I supposed to do?

Author: ```suggestion
Otherwise, you need to perform further checks to confirm the situation. Read on to learn
about these further checks.
```
?

Reviewer: I'd suggest we remove this paragraph and add one sentence after line 55:

If it is not specified, you do not have encryption at rest enabled.

If `--encryption-provider-config` is set, check which resources (such as `secrets`) are
configured for encryption, and what provider is used.
Make sure that the preferred provider for that resource type is **not** `identity`; you
only set `identity` (_no encryption_) as default when you want to disable encryption at
rest.
Verify that the first-listed provider for a resource is something **other** than `identity`,
which means that any new information written to resources of that type will be encrypted as
configured. If you do see `identity` as the first-listed provider for any resource, this

> means that those resources are being written out to etcd without encryption.

Reviewer: Does this paragraph have to be so wordy?

Author: This is the best wording I have capacity for. Follow up PRs are very welcome, so you could suggest different text that way.

> If `--encryption-provider-config` is set, check which resources (such as `secrets`) are
> configured for encryption, and what provider is used.
> Make sure that the preferred provider for that resource type is **not** `identity`; you
> only set `identity` (_no encryption_) as default when you want to disable encryption at
> rest.
> Verify that the first-listed provider for a resource is something **other** than `identity`,
> which means that any new information written to resources of that type will be encrypted as
> configured. If you do see `identity` as the first-listed provider for any resource, this
> means that those resources are being written out to etcd without encryption.

Reviewer: Please help check if the following revision is missing any points we want to make?

```suggestion
If `--encryption-provider-config` is specified, check which resources (such as `secrets`) are
configured for encryption, and which provider is used.
The first provider listed for a resource is the preferred one. If this provider is `identity`,
the resource is **not** encrypted, otherwise any new information written to the resources
of that type is encrypted using the specified provider.
```

Author: I think the new text covers this. Would be good to have a tech review confirm that.

{{< note >}}
It may not be enough to see that encryption at rest is active for new data.

If you are not sure about the progress of any previous migration to encrypt data at rest,
you must overwrite each resource with a new updated copy, in order to trigger encryption.
See [Ensure all relevant data are encrypted](#ensure-all-secrets-are-encrypted) for details.
{{< /note >}}

## Encryption-at-rest configuration {#understanding-the-encryption-at-rest-configuration}

{{< caution >}}
For cluster configurations with two or more control plane nodes, the encryption configuration
**must** be identical across each control plane node.

If there is a difference in the encryption provider configuration, this may well mean
that the kube-apiserver can't decrypt data stored inside the key-value store (potentially

> leading to further problems).

Reviewer: Line 90-92 further elaborates the "must", but I failed to see a need for this paragraph.
This can be easily deduced.
We can elaborate the negative effects of such inconsistent settings, but obviously we cannot enumerate all cases.

Author: When we tell people that something **must** happen, it's good to provide context. This lets the reader evaluate whether to heed the advice.

For example, sometimes we write a **must** but all that happens if you break it is that some invalid object gets autoreplaced, and your cluster carries on. In another scenario, none of your ServiceAccounts work any more, the control plane degrades, and this is overall a very bad outcome. Context does help.

---

<!-- overview -->
Some of the APIs in Kubernetes, such as {{< glossary_tooltip text="Secret" term_id="secret" >}},
support at-rest encryption. This at-rest encryption is additional to any system-level
encryption for the etcd cluster or hosts where the kube-apiserver stores data persistently.

This page shows how to enable and configure encryption of secret data at rest.

{{< note >}}
This task covers encryption for data stored using the
{{< glossary_tooltip text="Kubernetes API" term_id="kubernetes-api" >}}.

If you want to encrypt data in filesystems that are mounted into containers, you instead need
to use a storage integration that provides encrypted
{{< glossary_tooltip text="volumes" term_id="volume" >}}.
{{< /note >}}

## {{% heading "prerequisites" %}}

* {{< include "task-tutorial-prereqs.md" >}} {{< version-check >}}

* Your cluster's control plane must use etcd v3.x (major version 3, any minor version)

<!-- steps -->

## Configuration and determining whether encryption at rest is already enabled

> By default, the API server uses `identity` provider is used to protect secrets in etcd, which

Author: Hmm, wrong grammar here too.

provides no encryption.

The `kube-apiserver` process accepts an argument `--encryption-provider-config`
that controls how API data is encrypted in etcd.
The configuration is provided as an API named
[`EncryptionConfiguration`](/docs/reference/config-api/apiserver-encryption.v1/).
You can see an example configuration
in [Encryption at rest configuration](#understanding-the-encryption-at-rest-configuration).

{{< caution >}}

> For cluster configurations that use with two or more control plane nodes, the encryption configuration

Reviewer 2: ```suggestion
For cluster configurations that uses two or more control plane nodes, the encryption configuration
```
OR
```suggestion
For cluster configurations with two or more control plane nodes, the encryption configuration
```

---

<!-- overview -->

> Some of the APIs in Kubernetes, such as {{< glossary_tooltip text="Secret" term_id="secret" >}},
> support at-rest encryption. This at-rest encryption is additional to any system-level

Reviewer: I'm not sure if the encryption is supported on a per-API basis. The encryption configuration actually
allows us to choose which GVK to encrypt. It is okay to encrypt Deployments instead of Secrets.

```suggestion
Kubernetes support at-rest encryption for resources. The encryption is useful for some sensitive data
such as {{< glossary_tooltip text="Secret" term_id="secret" >}} objects.
This at-rest encryption is additional to any system-level
```

Author: You can't choose which resources are and aren't encrypted, not in the HTTP sense of resource (each resource in HTTP has a unique URI). You do indeed configure it at the group / kind level, which is what I mean by ”per API”.

Reviewer: Emm ... I'd disagree with you on the use of 'API' here. You can say you are encrypting an API resource, an API kind.
In this context, we are not talking about HTTP resources. K8s has a narrow definition of resources.

encryption for the etcd cluster or for the filesystem(s) on hosts where you are running the
kube-apiserver.

This page shows how to enable and configure encryption of API data at rest.

{{< note >}}
This task covers encryption for data stored using the
{{< glossary_tooltip text="Kubernetes API" term_id="kubernetes-api" >}}.

If you want to encrypt data in filesystems that are mounted into containers, you instead need
to either:

- use a storage integration that provides encrypted
{{< glossary_tooltip text="volumes" term_id="volume" >}}
- encrypt the data within your own application
{{< /note >}}

## {{% heading "prerequisites" %}}

* {{< include "task-tutorial-prereqs.md" >}}

* This task assumes that you are running the Kubernetes API server as a
  {{< glossary_tooltip text="static pod" term_id="static-pod" >}} on each control
  plane node.

* Your cluster's control plane **must** use etcd v3.x (major version 3, any minor version)

* To encrypt a custom resource, your cluster must be running Kubernetes v1.26 or newer.

<!-- steps -->

## Determine whether encryption at rest is already enabled {#determining-whether-encryption-at-rest-is-already-enabled}

> 

Reviewer 3: Since this is a *task* page, I'd like clear steps to complete some tasks.
Should this section list the items that you need to check before creating an encryption-at-rest configuration?

Not sure what to do using this sentence:
By default, the `identity` provider used by the API server to protect confidential data ...

Author: This page is in the tasks section but it is, unfortunately, also serving as a conceptual explanation. It would be nice to separate these things out; however, I think that's best saved for a separate PR.

> By default, the `identity` provider used by the API server to protect confidential data

Reviewer: ```suggestion
By default, the API server uses an `identity` provider to protect confidential data
```

[`EncryptionConfiguration`](/docs/reference/config-api/apiserver-encryption.v1/).
You can see an example configuration
in [Encryption at rest configuration](#understanding-the-encryption-at-rest-configuration).

Check whether the `kube-apiserver` process is running with the `--encryption-provider-config`
command line argument set to `true`. If it is not, you do not have encryption at rest enabled.

If `--encryption-provider-config` is set, check which resources (such as `secrets`) are
configured for encryption, and what provider is used.
Make sure that the preferred provider for that resource type is **not** `identity`; you
only set `identity` (_no encryption_) as default when you want to disable encryption at
rest.
Then the first-listed provider for a resource is something **other** than `identity`,
and any new information written to resources of that type will be encrypted as configured.

If you are not sure about the progress of any previous migration to encrypt data at rest,
you must overwrite each resource with a new updated copy, in order to trigger encryption.
See [Encrypting your data](#encrypting-your-data) for details.

> ## Encryption-at-rest configuration {#understanding-the-encryption-at-rest-configuration}

Reviewer 3: This section has useful information, but where does the content belong?

Author: Perhaps that could also happen in a follow-up PR? When I've incorporated wider feedback into PRs, I've found that they then end up at a size that puts people off reviewing them.

a previous migration to encrypted storage has succeeded. If you are not sure, see
[ensure all relevant data are encrypted](#ensure-all-secrets-are-encrypted).

## Plan an encryption-at-rest configuration {#understanding-the-encryption-at-rest-configuration}

{{< note >}}
For cluster configurations with two or more control plane nodes, the encryption configuration

> should be identical across each control plane node.

Author: This is repeated as a caution block later in the file (under the heading ”Reconfigure other control plane hosts”).

I think the advice bears repeating. Get this wrong and you can really end up regretting it.

`kms` | Uses envelope encryption scheme: Data is encrypted by data encryption keys (DEKs) using AES-CBC with [PKCS#7](https://datatracker.ietf.org/doc/html/rfc2315) padding; DEKs are encrypted by key encryption keys (KEKs) according to configuration in Key Management Service (KMS) | Strongest | Fast | 32-bytes | The recommended choice for using a third party tool for key management. Simplifies key rotation, with a new DEK generated for each encryption, and KEK rotation controlled by the user. See [Configure the KMS provider](/docs/tasks/administer-cluster/kms-provider/). May incur extra costs for the external key management.
`secretbox` | XSalsa20 and Poly1305 | Strong | Faster | 32-byte | Uses relatively new encryption technologies that may not be considered acceptable in environments that require high levels of review. Key material held within Kubernetes cluster.

You can configure multiple providers, and each provider (other than `identity`, which does not encrypt) supports multiple keys.
For encryption, the API server uses the first configured key from the first provider.

> For decryption, the API server tries each key in order for decryption, stopping when decryption succeeds.

Author: What's the different wording you're suggesting @divya-mohan0209 ?

> For decryption, the API server tries each key in order for decryption, stopping when decryption succeeds.

Reviewer 2: So, what I intended to ask was "For decryption, the API server tries each key in order for decryption, stopping when decryption succeeds." is the order sequential?

> For decryption, the API server tries each key in order for decryption, stopping when decryption succeeds.

Author: Hope I covered this.

              secret: YWJjZGVmZ2hpamtsbW5vcHFyc3R1dnd4eXoxMjM0NTY=
```

> Each `resources` array item is a separate configuration and contains a complete configuration. The

Reviewer 2: I know we're elaborating on the resources array item by speaking about individual fields. However, the first sentence feels incomplete to me. Perhaps we should include what it contains a complete configuration of?

`resources.resources` field is an array of Kubernetes resource types that the API server should
encrypt.
You can specify resource types as a simple type based on its URL representation (for example,
`secrets` for the [Secret API](/docs/reference/kubernetes-api/config-and-storage-resources/secret-v1/)).
You can also specify encryption for an API group (for example: `events.k8s.io`)
that should be encrypted. The `providers` array is an ordered list of the possible encryption

> providers. Only one provider type may be specified per top-level entry

Reviewer 2: Agree with Qiming.

> providers. Only one provider type may be specified per top-level entry

Author: I think the new wording covers this.

that the API server should encrypt.
You can specify resource types as a simple type based on its URL representation (for example,
`secrets` for the [Secret API](/docs/reference/kubernetes-api/config-and-storage-resources/secret-v1/)).
You can also specify encryption for an API group (for example: `events.k8s.io`)
that should be encrypted. The `providers` array is an ordered list of the possible encryption
providers. Only one provider type may be specified per top-level entry
(for example: you can specify `identity` or `aescbc`, but not both in the same item).

If custom resources are added to `EncryptionConfiguration` and the cluster version is 1.26 or newer,
any newly created custom resources mentioned in the `EncryptionConfiguration` will be encrypted.
Any custom resources that existed in etcd prior to that version and configuration will be unencrypted
until they are next written to storage. This is the same behavior as built-in resources.
See the [Ensure all secrets are encrypted](#ensure-all-secrets-are-encrypted) section.

Within an item, the `providers` array is an ordered list of the possible encryption providers to use
for the APIs that you listed.

> This ordered list can contain more than one provider - the order matters - but you cannot repeat the

Reviewer 3: nit: Are these `em` dashes or hyphens?

Author: Right now that's a hyphen. I could use an en dash or em dash if that's important.

same provider within one top-level item. For example, a top-level `resources` item can specify
a `providers` array with details for `aescbc` and for `identity`, in that order, but not for
`aescbc` and then a different `aescbc` configuration).

> For encryption, Kubernetes always uses the details you specify first. The other providers in the

Reviewer 3: nit: I did not understand what the `details` are
Could you clarify by instead using `providers`?

Author: `providers` would be wrong. It is in fact the first item from the `providers` array. I can reword.

same list are only used for decryption: Kubernetes attempts to decrypt the stored data using each
provider in turn, and stops if this succeeds.
If no provider can read the stored data due to a mismatch in format or secret key, an error
is returned, and that error prevents clients from accessing that resource.

{{< note >}}
If any resource is not readable via the encryption configuration (because keys were changed),
and you cannot restore a working configuration, your only recourse is to delete that entry from
the underlying etcd directly.

Any calls to the Kubernetes API that attempt to read that resource will fail until it is deleted
or a valid decryption key is provided.
{{< /note >}}

For more detailed information about the `EncryptionConfiguration` for the API server,
refer to the
[encryption configuration reference](/docs/reference/config-api/apiserver-encryption.v1/).

### Providers

{{< table caption="Providers for Kubernetes encryption at rest" >}}
Name | Encryption | Strength | Speed | Key length | Other considerations
-----|------------|----------|-------|------------|---------------------

> `identity` | **None** | N/A | N/A | N/A | Resources written as-is without encryption. When set as the first provider, the resource will be decrypted as new values are written. Existing encrypted resource are **not** automatically overwritten with the plaintext data.

Reviewer 3: nit: The first column in the table wraps the text (probably caused by the table style). I'd like to see the table data not wrapped.

Author: I think we have an issue logged about improving the rendering.

{{< /table >}}

By default, the `identity` provider is used to protect data in etcd, which provides **no**
encryption. `EncryptionConfiguration` was introduced to encrypt confidential data locally, using
a locally managed key.

Encrypting secret data with a locally managed key protects against an etcd compromise, but it fails to
protect against a host compromise. Since the encryption keys are readable - in the clear - on the
control plane host in the EncryptionConfiguration YAML file, a skilled attacker can access that file
and extract the encryption keys.

Envelope encryption creates dependence on a separate key, not stored in Kubernetes. In this case,
an attacker would need to compromise etcd, the `kubeapi-server`, and the third-party KMS provider to
retrieve the plaintext values, providing a higher level of security than locally stored encryption keys.

> ## Encrypting your data

Reviewer 3: Possibly change heading to
Encrypt your data

Sometimes the main headings convey an action, sometimes a concept

on the (external) KMS integration that you are using.

### Generate the encryption key

Start by generating a new encryption key, and then encode it using base64:

{{< tabs name="generate_encryption_key" >}}
{{% tab name="Linux" %}}
Generate a 32-byte random key and base64 encode it. You can use this command:
```shell
head -c 32 /dev/urandom | base64
```

You can use `/dev/hwrng` instead of `/dev/urandom` if you want to
use your PC's built-in hardware entropy source. Not all Linux
devices provide a hardware random generator.
{{% /tab %}}
{{% tab name="macOS" %}}
<!-- localization note: this is similar to the Linux tab and the wording
should match wherever the English text does -->
Generate a 32-byte random key and base64 encode it. You can use this command:
```shell
head -c 32 /dev/urandom | base64
```
{{% /tab %}}
{{% tab name="Windows" %}}
Generate a 32-byte random key and base64 encode it. You can use this command:
```powershell
# Do not run this in a session where you have set a random number
# generator seed.
[Convert]::ToBase64String((1..32|%{[byte](Get-Random -Max 256)}))
```
{{% /tab %}}
{{< /tabs >}}

{{< note >}}

> Keep the encryption secret confidential, including whilst you generate it and

Reviewer 3: nit: Keep the encryption secret confidential, including while ...

Reviewer: Let's call it a key, for consistency.
In this page, I'm always thinking of 'secret' as the thing to be encrypted.

{{< note >}}
If any resource is not readable via the encryption configuration (because keys were changed),
and you cannot restore a working configuration, your only recourse is to delete that entry from
the underlying etcd directly.

Any calls to the Kubernetes API that attempt to read that resource will fail until it is deleted
or a valid decryption key is provided.
{{< /note >}}

## Providers

{{< table caption="Providers for Kubernetes encryption at rest" >}}
Name | Encryption | Strength | Speed | Key length | Other considerations
-----|------------|----------|-------|------------|---------------------
`identity` | **None** | N/A | N/A | N/A | Resources written as-is without encryption. When set as the first provider, the resource will be decrypted as new values are written. Existing encrypted resource are **not** automatically overwritten with the plaintext data.
`aescbc` | AES-CBC with [PKCS#7](https://datatracker.ietf.org/doc/html/rfc2315) padding | Weak | Fast | 32-byte | Not recommended due to CBC's vulnerability to padding oracle attacks. Key material held within Kubernetes cluster.

> `aesgcm` | AES-GCM with random nonce | Must be rotated every 200,000 writes | Fastest | 16, 24, or 32-byte | Not recommended for use except when an automated key rotation scheme is implemented. Key material held within Kubernetes cluster.

Author: It means that within the Kubernetes cluster itself, the data of the encryption key are stored there. That approach is different from the `kms` mechanism where the key material is held in a service that runs outside the cluster.

`kms` | Uses envelope encryption scheme: Data is encrypted by data encryption keys (DEKs) using AES-CBC with [PKCS#7](https://datatracker.ietf.org/doc/html/rfc2315) padding; DEKs are encrypted by key encryption keys (KEKs) according to configuration in Key Management Service (KMS) | Strongest | Fast | 32-bytes | The recommended choice for using a third party tool for key management. Simplifies key rotation, with a new DEK generated for each encryption, and KEK rotation controlled by the user. See [Configure the KMS provider](/docs/tasks/administer-cluster/kms-provider/). May incur extra costs for the external key management.
`secretbox` | XSalsa20 and Poly1305 | Strong | Faster | 32-byte | Uses relatively new encryption technologies that may not be considered acceptable in environments that require high levels of review. Key material held within Kubernetes cluster.
{{< /table >}}

> You can configure multiple providers, and each provider (other than `identity`, which does not encrypt) supports multiple keys.

Author: Is this repetition harmful? I think it may help the reader.

providers. Only one provider type may be specified per top-level entry
(for example: you can specify `identity` or `aescbc`, but not both in the same item).

If custom resources are added to `EncryptionConfiguration` and the cluster version is 1.26 or newer,
any newly created custom resources mentioned in the `EncryptionConfiguration` will be encrypted.
Any custom resources that existed in etcd prior to that version and configuration will be unencrypted
until they are next written to storage. This is the same behavior as built-in resources.
See the [Ensure all secrets are encrypted](#ensure-all-secrets-are-encrypted) section.

Within an item, the `providers` array is an ordered list of the possible encryption providers to use
for the APIs that you listed.

This ordered list can contain more than one provider - the order matters - but you cannot repeat the
same provider within one top-level item. For example, a top-level `resources` item can specify
a `providers` array with details for `aescbc` and for `identity`, in that order, but not for
`aescbc` and then a different `aescbc` configuration).

> For encryption, Kubernetes always uses the first item from the `providers` array. The other entries

Author: In tech docs, repetition can have value. The reader may not have worked it out from the first mention.

I think this detail bears repeating.

in that configuration array are used for decryption: Kubernetes attempts to decrypt the stored data
using each provider in turn, starting at the first entry that is also used for encryption.
If decryption succeeds, Kubernetes uses that data.
If no provider can read the stored data (perhaps due to a mismatch in format or secret key),
an error is returned, and that error prevents clients from accessing that resource.

{{< caution >}}
If any resource is not readable via the encryption configuration (because keys were changed),
and you cannot restore a working configuration, your only recourse is to delete that entry from
the underlying etcd directly.

Any calls to the Kubernetes API that attempt to read that resource will fail until it is deleted

> or a valid decryption key is provided.

Author: In the case that you **can** restore a working configuration, the failures stop. I don't see a conflict.

{{< /table >}}

By default, the `identity` provider is used to protect data in etcd, which provides **no**
encryption. `EncryptionConfiguration` was introduced to encrypt confidential data locally, using
a locally managed key.

Encrypting secret data with a locally managed key protects against an etcd compromise, but it fails to
protect against a host compromise. Since the encryption keys are readable - in the clear - on the
control plane host in the EncryptionConfiguration YAML file, a skilled attacker can access that file
and extract the encryption keys.

Envelope encryption creates dependence on a separate key, not stored in Kubernetes. In this case,
an attacker would need to compromise etcd, the `kubeapi-server`, and the third-party KMS provider to
retrieve the plaintext values, providing a higher level of security than locally stored encryption keys.

## Encrypt your data {#encrypting-your-data}

{{< caution >}}
Storing the raw encryption key in the EncryptionConfig only moderately improves your security posture,

> compared to no encryption.

Author: It can be OK to make the same point more than once in a document. See (eg) [The Use of Repetition in Technical Communication](https://kmh-lanl.hansonhub.com/pc-26-9-zimmerman.pdf).

I think this is one of those occasions. If someone has skipped to this heading, it's still worth showing them this admonition.

```
{{% /tab %}}
{{% tab name="Windows" %}}
Generate a 32-byte random key and base64 encode it. You can use this command:
```powershell
# Do not run this in a session where you have set a random number
# generator seed.
[Convert]::ToBase64String((1..32|%{[byte](Get-Random -Max 256)}))
```
{{% /tab %}}
{{< /tabs >}}

{{< note >}}
Keep the encryption secret confidential, including whilst you generate it and
ideally even after you are no longer actively using it.
{{< /note >}}

### Replicate the encryption key

Using a secure mechanism for file transfer, make a copy of that encryption key
available to every other control plane host.

At a minimum, use encryption in transit - for example, secure shell (SSH). For more
security, use asymmetric encryption between hosts, or change the approach you are using
so that you're relying on KMS encryption.

### Write an encryption configuration file

{{< caution >}}
The encryption configuration file may contain keys that can decrypt content in etcd.
If the configuration file contains any key material, you must properly
restrict permissions on all your control plane hosts so only the user
who runs the kube-apiserver can read this configuration.

> You may need to create a secure backup of the configuration file. If so, do

Reviewer: Which "configuration file" should I backup? The encryption configuration file is yet to be written.

an attacker would need to compromise etcd, the `kubeapi-server`, and the third-party KMS provider to
retrieve the plaintext values, providing a higher level of security than locally stored encryption keys.

{{< caution >}}
Storing the raw encryption key in the EncryptionConfig only moderately improves your security posture, compared to no encryption.

> For additional secrecy, consider using the `kms` provider which relies on keys held outside your Kubernetes cluster.

Reviewer 2: I like this suggestion.

> For additional secrecy, consider using the `kms` provider which relies on keys held outside your Kubernetes cluster.

Author: The `kms` provider doesn't _manage_ those external keys; it consumes a service that provides access to the managed keys. I will consider how to reword.

      - aescbc:
          keys:
            - name: key1
              # See the following text for more details about the secret value
              secret: <BASE 64 ENCODED SECRET>
      - identity: {} # this fallback allows reading unencrypted secrets;
                     # for example, during initial migratoin
```

Place that base64 encoded value in the `secret` field of the encryption configuration file.

### Reconfigure the API server

{{< caution >}}
The encryption configuration file may contain keys that can decrypt content in etcd.
If the configuration file contains any key material, you must properly
restrict permissions on all your control plane hosts so only the user
who runs the kube-apiserver can read this configuration.

Consider whether you need to make a backup of that configuration. If you do,

> also think about how you will ensure that your backup is secure.

Reviewer: I'd prefer reading simple clear text over this. I mean:

```
You may want to create a secure backup of the configuration file.
```

   control plane hosts (strictly speaking: on all hosts that run a `kube-apiserver`).
2. Restart all `kube-apiserver` processes, to ensure each control plane host can decrypt
   any data that are encrypted with the new key
   (this is required for a replicated control plane; you can skip this step if you only run a
   single `kube-apiserver`).
3. Make a secure backup of the new encryption key. If you lose all copies of this key you would
   need to delete all the resources were encrypted under the lost key, and workloads may not
   operate as expected during the time that at-rest encryption is broken.
4. Make the new key the first entry in the `keys` array so that it is used for encryption-at-rest
   for new writes
5. Restart all `kube-apiserver` processes to ensure each control plane host now encrypts using the new key
6. As a privileged user, run `kubectl get secrets --all-namespaces -o json | kubectl replace -f -`
   to encrypt all existing Secrets with the new key
7. After you have updated all existing Secrets to use the new key and have made a secure backup of the
   new key, remove the old decryption key from the configuration.

> When your cluster only has a single `kube-apiserver` instance, you can skip step 2.

Reviewer 2: I think this is redundant? It's already mentioned after step #2?

## Decrypting all data

To disable encryption at rest, place the `identity` provider as the first

> entry your encryption configuration file:

Reviewer 2: ```suggestion
entry in your encryption configuration file:
```

kubectl get secrets --all-namespaces -o json | kubectl replace -f -
```

The command above reads all Secrets and then updates them with the same data, in order to
apply server side encryption.

{{< note >}}
If an error occurs due to a conflicting write, retry the command.
It is safe to run that command more than once.

For larger clusters, you may wish to subdivide the Secrets by namespace,
or script an update.
{{< /note >}}

## Cleanup {#cleanup-all-secrets-encrypted}

You **must** ensure that all existing Secrets are encrypted at rest before
performing this cleanup step, because this change prevents the API server from
retrieving Secret resources (API path `secrets`) that are not encrypted.

Once all Secrets in your cluster are encrypted, you can remove the `identity`
part of the encryption configuration. For example:

```yaml
---
apiVersion: apiserver.config.k8s.io/v1
kind: EncryptionConfiguration
resources:
  - resources:
      - secrets
    providers:
      - aescbc:
          keys:
            - name: key1
              secret: <BASE 64 ENCODED SECRET>
      - identity: {} # REMOVE THIS LINE
```

…and then restart each API server in turn. This change prevents the API server

> from accessing a plain-text Secret, even by accident.

Reviewer: This may and may not be a good practice. It depends.
If I'm using one provider today and I'm migrating to a different provider tomorrow, keeping `identity` there may be convenient. I don't know. My gut feeling is that an assertion like this should be made carefully.

Author: When is it not a good practice to remove the `identity` provider?

## Rotating a decryption key

Changing a Secret without incurring downtime requires a multi-step operation, especially in
the presence of a highly-available deployment where multiple `kube-apiserver` processes are running.

1. Generate a new key and add it as the second key entry for the current provider on all

>    control plane hosts (strictly speaking: on all hosts that run a `kube-apiserver`).

Reviewer: ```suggestion
   nodes that run a `kube-apiserver` process.
```

Author: My wording avoids assuming that the `kube-apiserver` is running inside Pods. I will try to find another wording that also works for the common case where there are control plane nodes.

> 1. Restart all `kube-apiserver` processes, to ensure each control plane host can decrypt
>    any data that are encrypted with the new key
>    (this is required for a replicated control plane; you can skip this step if you only run a
>    single `kube-apiserver`).

Reviewer: ```suggestion
1. Restart **all** `kube-apiserver` processes, to ensure each server can decrypt
   any data that are encrypted with the new key.
```

      - aescbc:
          keys:
            - name: key1
              # See the following text for more details about the secret value
              secret: <BASE 64 ENCODED SECRET>
      - identity: {} # this fallback allows reading unencrypted secrets;
                     # for example, during initial migratoin
```

Place that base64 encoded value in the `secret` field of the encryption configuration file.

### Update the configuration of each API server {#api-server-config-update}

> You will need to mount the new encryption configuration file to the `kube-apiserver`

Reviewer 3: nit: Could shorten to,
Mount the new encryption configuration ...

Author: I'll leave this as I'm not rewording here.

     - command:
       - kube-apiserver
       ...
       - --encryption-provider-config=/etc/kubernetes/enc/enc.yaml  # add this line
       volumeMounts:
       ...
       - name: enc                           # add this line
         mountPath: /etc/kubernetes/enc      # add this line
         readonly: true                      # add this line
       ...
     volumes:
     ...
     - name: enc                             # add this line
       hostPath:                             # add this line
         path: /etc/kubernetes/enc           # add this line
         type: DirectoryOrCreate             # add this line
     ...
   ```

> 1. Restart the API server; for example, using `crictl stopp`.

Reviewer: No experience on `crictl`, is `crictl stopp` a typo? Please confirm.

Reviewer: Actually, we don't need to do this. The static pod will be recreated if the YAML file is changed. Right?

Author: `stopp` is not a typo.

We don't know how often the kubelet polls for static pod manifests. The default sync frequency, 60s, is not too bad, but it might be quite a bit higher.

   (When you manually stop a static Pod, the kubelet automatically creates a new
    replacement Pod based on the current manifest.)

### Reconfigure other control plane hosts {#api-server-config-update-more}

If you have multiple API servers in your cluster, you should deploy the
changes in turn to each API server.

Make sure that you use the **same** encryption configuration on each
control plane host.

### Verify that newly written data is encrypted {#verifying-that-data-is-encrypted}

> Data is encrypted when written to etcd. After restarting your `kube-apiserver`, any newly created or updated Secret (or other resource types configured in

Reviewer: The existing line wrapping is good.

   ```

   The output should contain `mykey: bXlkYXRh`, with contents of `mydata` encoded using base64;
   read
   [decoding a Secret](/docs/tasks/configmap-secret/managing-secret-using-kubectl/#decoding-secret)
   to learn how to completely decode the Secret.

### Ensure all relevant data are encrypted {#ensure-all-secrets-are-encrypted}

For this example, you have configured your cluster so that Secrets are encrypted on write.
Performing a replace operation for each Secret will encrypt that content at rest,
where the objects are unchanged.

You can make this change across all Secrets in your cluster:

```shell
# Run this as an administrator that can read and write all Secrets
kubectl get secrets --all-namespaces -o json | kubectl replace -f -
```

The command above reads all Secrets and then updates them with the same data, in order to
apply server side encryption.

{{< note >}}
If an error occurs due to a conflicting write, retry the command.
It is safe to run that command more than once.

For larger clusters, you may wish to subdivide the Secrets by namespace,
or script an update.
{{< /note >}}

## Cleanup {#cleanup-all-secrets-encrypted}

{{< warning >}}

> Making this change prevents the API server from retrieving resources that are marked

Reviewer 3: nit: Clean up <what>
nit: What is the change that the reader is making?

Author: It is: remove the `identity` part of the encryption configuration. The heading could be more obvious.

Author: How's the new wording here?

encryption, you can remove the API server's ability to read that API's backing data
as plaintext.

{{< warning >}}
Making this change prevents the API server from retrieving resources that are marked
as encrypted as rest, but are actually stored in the clear.

When you have configured encryption at rest for an API (for example: the API kind
`Secret`, representing `secrets` resources in the core API group), you **must** ensure
that all those resources in this cluster really are encrypted at rest. Check this before
you carry on with the next steps.
{{< /warning >}}

Once all Secrets in your cluster are encrypted, you can remove the `identity`
part of the encryption configuration. For example:

```yaml
---
apiVersion: apiserver.config.k8s.io/v1
kind: EncryptionConfiguration
resources:
  - resources:
      - secrets
    providers:
      - aescbc:
          keys:
            - name: key1
              secret: <BASE 64 ENCODED SECRET>
      - identity: {} # REMOVE THIS LINE
```

…and then restart each API server in turn. This change prevents the API server
from accessing a plain-text Secret, even by accident.

## Rotate a decryption key {#rotating-a-decryption-key}

> Changing an encryption key for Kubernetes without incurring downtime requires a multi-step operation,

Author: I felt that the heading “Rotating a decryption key” was unambiguously about a scenario where objects were encrypted with key _A_ and you wanted to change to have them encrypted under key _B_.

I still think that's right.

---

Where did the existing page talk about migrating from unencrypted to encrypted data? There is https://kubernetes.io/docs/tasks/administer-cluster/encrypt-data/#ensure-all-secrets-are-encrypted which is represented by the text following
```markdown
### Ensure all relevant data are encrypted {#ensure-all-secrets-are-encrypted}
```
in this revised document.

resources:
  - resources:
      - secrets
      # list any other resources here that you previously were
      # encrypting at rest
    providers:
      - identity: {} # add this line
      - aescbc:
          keys:
            - name: key1
              secret: <BASE 64 ENCODED SECRET> # keep this in place
                                               # make sure it comes after "identity"
```

Then run the following command to force decryption of all Secrets:

```shell
kubectl get secrets --all-namespaces -o json | kubectl replace -f -
```

Once you have replaced all existing encrypted resources with backing data that
don't use encryption, you can remove the encryption settings from the
`kube-apiserver`.

## Configure automatic reloading

You can specify a boolean `--encryption-provider-config-automatic-reload` command line
argument to `kube-apiserver`. That setting determines whether the API server should automatically
reload the file you specify for `--encryption-provider-config`. Enabling this option allows you
to change the keys for encryption at rest without restarting the API server.

To allow automatic reloading, configure the API server to run with:

> `--encryption-provider-config-automatic-reload=true`

Reviewer: This is again redundant.

Author: I disagree, because we haven't told readers that it's an argument that takes a boolean value. The reader can infer it, but in tech docs it pays to be precise.

OK, OK, it doesn't pay very well if your an open source contributor - but I hope you get my point.

    replacement Pod based on the current manifest.)

{{< note >}}
Depending on your information security needs, you may want to create a secure backup of the
configuration file. If appropriate, do that before you continue.

If you do take a backup, it's good practice to check you can read it afterwards.
{{< /note >}}

You now have encryption in place for **one** control plane host. A typical
Kubernetes cluster has multiple control plane hosts, so there is more to do.

#### Reconfigure other control plane hosts {#api-server-config-update-more}

If you have multiple API servers in your cluster, you should deploy the
changes in turn to each API server.

{{< caution >}}
For cluster configurations with two or more control plane nodes, the encryption configuration
should be identical across each control plane node.

If there is a difference in the encryption provider configuration between control plane
nodes, this difference may mean that the kube-apiserver can't decrypt data.
{{< /caution >}}

When you are planning to update the encryption configuration of your cluster, plan this
so that the API servers in your control plane can always decrypt the stored data

> (even part way through rolling out the change).

Author: This advice is new as of today.
